pkg/config/structure/unmarshal_test.go

```diff
@@ -814,6 +784,443 @@ feature:
 	assert.NoError(t, err)
 }
 
+func TestUnmarshalKeyConversionErrors(t *testing.T) {
+	t.Run("errors on string to int", func(t *testing.T) {
+		confYaml := `
+feature:
+  enabled: "true"
+`
+
+		mockConfig := mock.NewFromYAML(t, confYaml)
+		mockConfig.SetKnown("feature")
+
+		feature := struct {
+			Enabled int
+		}{}
+
+		err := UnmarshalKey(mockConfig, "feature", &feature)
+		assert.Error(t, err)
+		assert.Contains(t, err.Error(), "could not convert to int")
+	})
+
+	t.Run("errors on string to float", func(t *testing.T) {
+		confYaml := `
+feature:
+  enabled: "true"
+`
+
+		mockConfig := mock.NewFromYAML(t, confYaml)
+		mockConfig.SetKnown("feature")
+
+		feature := struct {
+			Enabled float64
+		}{}
+
+		err := UnmarshalKey(mockConfig, "feature", &feature)
+		assert.Error(t, err)
+		assert.Contains(t, err.Error(), "could not convert to float")
+	})
+
+	t.Run("errors on float to bool", func(t *testing.T) {
+		confYaml := `
+feature:
+  enabled: 1.0
+`
+
+		mockConfig := mock.NewFromYAML(t, confYaml)
+		mockConfig.SetKnown("feature")
+
+		feature := struct {
+			Enabled bool
+		}{}
+
+		err := UnmarshalKey(mockConfig, "feature", &feature)
+		assert.Error(t, err)
+		assert.Contains(t, err.Error(), "could not convert to bool")
+	})
+
+	t.Run("errors on negative to uint", func(t *testing.T) {
+		t.Skip("not implemented")
+		confYaml := `
+feature:
+  enabled: -1
```

Comment:
it looks like [viper will error on negative values](https://github.com/spf13/cast/blob/master/caste.go#L540-L547) when casting to uints, we don't currently and also don't underflow

Reply:
Good to have this bug in the new implementation documented, it will be fixed in a follow-up PR in order to get this test passing.